src/recorder.rs

```diff
@@ -0,0 +1,71 @@
+//! Parser extensions that inspect the input without modifying it.
+//!
+//! *"Only one man stood and watched the sky, stood with terrible sadness in his eyes
+//! and rubber bungs in his ears. He knew exactly what was happening and had known
+//! ever since his Sub-Etha Sens-O-Matic had started winking in the dead of night
+//! beside his pillar and woken him with a start."*
+use crate::{input::Marker, Input};
+use core::ops::{Deref, DerefMut};
+
+#[allow(unused)] // for intra-doc links
+use crate::Parser;
+
+/// A type that receives event hooks when certain parsing actions occur.
+///
+/// Unlike [`Parser::State`], this receives information about the incoming actions.
+///
+/// If you don't need to receive event hooks, use [`SimpleState`].
+pub trait Recorder<'a, I: Input<'a>> {
+    /// A type the Recorder can use to revert to a previous state.
+    ///
+    /// For implementation reasons, this is required to be `Copy + Clone`.
+    type SaveMarker: Copy + Clone;
+
+    /// This function is called when a new token is read from the input stream.
+    fn on_token(&mut self, token: I::Token);
```

Comment:
Is this ever getting invoked? I assume not, since this sadly doesn't generalise to by-ref and by-value inputs. The correct thing would be `token: I::TokenMaybe`.

Reply:
FWIW, the place to call it is in these functions:

- `InputRef::skip_while`
- `InputRef::next_inner`
- `InputRef::next_maybe_inner`
- `InputRef::next_ref_inner`

Reply:
it was really a pain to convert back and forth between `MaybeRef` :sweat_smile: i added those calls and changed the signature to pass in `&Token` instead.